### ISSUE TITLE
Middleware exception in AuthShopify

### ISSUE THREAD

**Issue by rusrocket** [osiset/laravel-shopify#593]
*For bug reporting only! If you're posting a feature request or discussion, please ignore.*

## Expected Behavior

Do not get this in the log

## Current Behavior

In laravel log I get: Array to string conversion {"exception":"[object] (ErrorException(code: 0): Array to string conversion at xxxxxxxx/vendor/osiset/laravel-shopify/src/ShopifyApp/Http/Middleware/AuthShopify.php:246)

## Failure Information

This error exist when login into shopify embedded app

**Comment by osiset** [comment]
Can you give some step by steps?

Line it's complaining about:

```php
$verify[$key] = is_array($value) ? '["'.implode('", "', $value).'"]' : $value;
```

Would you be able to log the incoming request URLs as well? I'd like to see if a query param is causing this.

**Comment by prashantbw** [comment]
Hi @osiset and @rusrocket 

  I was facing the same issue while using the datatables,
Cause:   getData function was receiving multi dimensional array so implode function was throwing an error

Solution:  
 if(is_array($value)&&$this->is_multi_array($value)){
                       $verify[$key]=implode(', ', array_map(function ($entry) {
                                          return is_array($entry) ? '["'.implode('", "', $entry).'"]' : $entry;
                                     }, $value));
                     }else{
                      $verify[$key] = is_array($value) ? '["'.implode('", "', $value).'"]' : $value;
                    }


hope that helps @rusrocket 

The condition can be finetuned FYI @osiset  Thanks

**Comment by osiset** [comment]
@prashantbw Whats is_multi_array ?

**Comment by mddickens** [comment]
I am getting these errors as well and at least in my case, they are caused by hacking attempts that contain the multi-dimensional arrays. Here's what I got in the log last night when I logged the contents of $request->all():

```
[2020-11-07 04:23:28] local.INFO: array (
  's' => '/Index/\\think\\app/invokefunction',
  'function' => 'call_user_func_array',
  'vars' => 
  array (
    0 => 'md5',
    1 => 
    array (
      0 => 'HelloThinkPHP21',
    ),
  ),
)  
```
I got some other junk as well but it's basically the same stuff

**Comment by osiset** [comment]
So this issue is to do with multi dimensional arrays in the params? Does anything do this in Shopify right now, I thought it was all flat arrays?

**Comment by mddickens** [comment]
It should be the usual 4 params (hmac, shopdomain, etc), but this is coming from outside of Shopify. I guess it some sot of crude hacking attempt? Anyway, I put the following at the top of getData in authShopify:

```
		if (count($request->all()) > 0 && !$request->has('hmac')) {
			throw new SignatureVerificationException('Invalid request');
		}
```

Not very sophisticated code, but it seems to work. I catch the exception and return a 400 to them without logging. FWIW

**Comment by NESTEROVILYASERGEEVICH** [comment]
> So this issue is to do with multi dimensional arrays in the params? Does anything do this in Shopify right now, I thought it was all flat arrays?

Hi there, I need your help with your package.
Could you please help me?

**Comment by prashantbw** [comment]
> @prashantbw Whats is_multi_array ?

@osiset a custom function written by me just to check if it is a multidimensional array

**Comment by osiset** [comment]
Internally utilized a similar function for a fix on #724

```php
    /**
     * Parse the data source value.
     * Handle simple key/values, arrays, and nested arrays.
     *
     * @param mixed $value
     *
     * @return string
     */
    private function parseDataSourceValue($value): string
    {
        /**
         * Format the value.
         *
         * @param mixed $val
         *
         * @return string
         */
        $formatValue = function ($val): string {
            return is_array($val) ? '["'.implode('", "', $val).'"]' : $val;
        };

        // Nested array
        if (is_array($value) && is_array(current($value))) {
            return implode(', ', array_map($formatValue, $value));
        }

        // Array or basic value
        return $formatValue($value);
    }
```

Awaiting physical tests, units pass right now.

**Comment by osiset** [comment]
Cleanup... Closing for now, cookies are removed in v17.0.0 and this middleware no longer exists in the same manner.
Re-open if still and issue.
Thanks!
